ParallelPrimitives/RadixSortKernels.h

```diff
@@ -521,278 +514,708 @@ extern "C" __global__ void SortSinglePassKernel( int* gSrcKey, int* gDstKey, int
+//	{
+//		for( int i = threadIdx.x; i < BIN_SIZE; i += SORT_WG_SIZE )
+//		{
+//			matchMasks[w][i] = 0;
+//		}
+//	}
+//
+//	__syncthreads();
+//
+//	for( int i = threadIdx.x; i < gNItemsPerWG; i += SORT_WG_SIZE )
+//	{
+//		const u32 itemIndex = blockIdx.x * gNItemsPerWG + i;
+//		if( itemIndex < numberOfInputs )
+//		{
+//			const auto item = gSrcKey[itemIndex];
+//			const u32 bucketIndex = getMaskedBits( item, START_BIT );
+//			atomicInc( &localPrefixSum[bucketIndex], 0xFFFFFFFF );
+//		}
+//	}
+//
+//	__syncthreads();
+//
+//	// Compute Prefix Sum
+//
+//	ldsScanExclusive( localPrefixSum, BIN_SIZE );
+//
+//	__syncthreads();
+//
+//	// Reorder
+//
+//	for( int i = threadIdx.x; i < gNItemsPerWG; i += SORT_WG_SIZE )
+//	{
+//		const u32 itemIndex = blockIdx.x * gNItemsPerWG + i;
+//
+//		const auto item = gSrcKey[itemIndex];
+//		const u32 bucketIndex = getMaskedBits( item, START_BIT );
+//
+//		const int warp = threadIdx.x / 32;
+//		const int lane = threadIdx.x % 32;
+//
+//		__syncthreads();
+//
+//		if( itemIndex < numberOfInputs )
+//		{
+//			atomicOr( &matchMasks[warp][bucketIndex], 1u << lane );
+//		}
+//
+//		__syncthreads();
+//
+//		bool flushMask = false;
+//
+//		u32 localOffset = 0;
+//		u32 localSrcIndex = 0;
+//
+//		if( itemIndex < numberOfInputs )
+//		{
+//			const u32 matchMask = matchMasks[warp][bucketIndex];
+//			const u32 lowerMask = ( 1u << lane ) - 1;
+//			u32 offset = __popc( matchMask & lowerMask );
+//
+//			flushMask = ( offset == 0 );
+//
+//			for( int w = 0; w < warp; ++w )
+//			{
+//				offset += __popc( matchMasks[w][bucketIndex] );
+//			}
+//
+//			localOffset = counters[bucketIndex] + offset;
+//			localSrcIndex = i;
+//		}
+//
+//		__syncthreads();
+//
+//		if( itemIndex < numberOfInputs )
+//		{
+//			atomicInc( &counters[bucketIndex], 0xFFFFFFFF );
+//		}
+//
+//		if( flushMask )
+//		{
+//			matchMasks[warp][bucketIndex] = 0;
+//		}
+//
+//		// Swap
+//
+//		if( itemIndex < numberOfInputs )
+//		{
+//			const u32 srcIndex = blockIdx.x * gNItemsPerWG + localSrcIndex;
+//			const u32 dstIndex = globalOffset[bucketIndex] + localOffset;
+//			gDstKey[dstIndex] = gSrcKey[srcIndex];
+//
+//			if constexpr( KEY_VALUE_PAIR )
+//			{
+//				gDstVal[dstIndex] = gSrcVal[srcIndex];
+//			}
+//		}
+//	}
+//}
+//
+//extern "C" __global__ void SortKernel( int* gSrcKey, int* gDstKey, int* gHistogram, int gN, int gNItemsPerWG, const int START_BIT, const int N_WGS_EXECUTED )
+//{
+//	SortImpl<false>( gSrcKey, nullptr, gDstKey, nullptr, gHistogram, gN, gNItemsPerWG, START_BIT, N_WGS_EXECUTED );
+//}
+//
+//extern "C" __global__ void SortKVKernel( int* gSrcKey, int* gSrcVal, int* gDstKey, int* gDstVal, int* gHistogram, int gN, int gNItemsPerWG, const int START_BIT, const int N_WGS_EXECUTED )
+//{
+//	SortImpl<true>( gSrcKey, gSrcVal, gDstKey, gDstVal, gHistogram, gN, gNItemsPerWG, START_BIT, N_WGS_EXECUTED );
+//}
+
+#define RADIX_SORT_KEY_TYPE u32
+#define RADIX_SORT_VALUE_TYPE u32
+#define KEY_IS_16BYTE_ALIGNED 1
```

Comment:
It's up to us. If we assume input and output pointers are 16-byte alignment, we can use the faster code path. Otherwise, some performance will be sacrificed